### PR TITLE
test: verify process instance modification in multi instance subprocesses using ancestor selection

### LINKED ISSUE #40226: [PoC] Modify elements in one instance of a multi-instance subprocess
### 

### Description

With this PoC, we want to prove that it's possible to elements inside an instance of a multi-instance subprocess body

### Acceptance criteria

- [ ] ...

### Links

_No response_

### Breakdown

_No response_

### PR DESCRIPTION
**Pull request by mustafadagher**
## Description

This PR refactors an existing test and adds three new test cases for process instance modification within multi-instance ad-hoc subprocesses. The changes improve test clarity by explicitly retrieving multi-instance subprocess keys rather than deriving them from task flow scopes.

- Refactored `shouldUseAncestorSelectionInsideMultiInstances` to explicitly retrieve multi-instance subprocess instances
- Added comprehensive test coverage for ancestor selection in ad-hoc subprocess scenarios

## Checklist

- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #40226

### CODE REVIEW

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 1]
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
...
+            .withVariables(Map.of("activateElements", List.of("A", "A")))
+            .create();
+
+    // Get all multi-instance ad-hoc subprocess instances
+    final var miInnerInstances =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementType(BpmnElementType.AD_HOC_SUB_PROCESS_INNER_INSTANCE)
+            .limit(2)
+            .toList();
```
Comment (mustafadagher) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: 💭 This here shows a problem that I don't know if it's intentional or not.
if this code block was replaced by a query for the inner `AD_HOC_SUB_PROCESS` instances and not THEIR INNER INSTANCES (`AD_HOC_SUB_PROCESS_INNER_INSTANCE`), modification would create a new `AD_HOC_SUB_PROCESS_INNER_INSTANCE` not attach the activated element to an existing inner instance.

This behavior is slightly different than `normal` (i.e not `ad-hoc`) subprocess body, where selecting the inner `SUB_PROCESS` is enough, and there is no `Inner_Instance` concept.  (see test `shouldUseAncestorSelectionInsideMultiInstances`)

```
final var subprocessBodyInstance =
        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
            .withProcessInstanceKey(processInstanceKey)
            .withElementId("adhoc_subprocess")
            .withElementType(BpmnElementType.MULTI_INSTANCE_BODY)
            .getFirst();

final var miInnerInstances =
        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
            .withProcessInstanceKey(processInstanceKey)
            .withElementType(BpmnElementType.AD_HOC_SUB_PROCESS)
            .withFlowScopeKey(subprocessBodyInstance.getKey())
            .limit(2)
            .toList();
```

---

💭 I have a test case locally that proves that a new `AD_HOC_SUB_PROCESS_INNER_INSTANCE` will be created in the scenario described above, ❓ just not sure if it's worth pushing here

--- 
💭 I also noticed that if we activate multiple elements inside an `ad-hoc` subprocess, each will have its own `AD_HOC_SUB_PROCESS_INNER_INSTANCE` created as a `flow scope` element. Meaning: in a single ad-hoc subprocess body, if we have 2 elements (`A`, `B`) and both are activated, there will be 2 separate `AD_HOC_SUB_PROCESS_INNER_INSTANCE`s created, one for each element.
❓ Question is, if this had an element `c` that was not active, what should be the expected behavior when modifying (activating) that `c` element? 
- if the above behavior is intended, then I assume creating a new `AD_HOC_SUB_PROCESS_INNER_INSTANCE` is the way to go.

Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: @remcowesterhoud maybe you know more about this.
❓ could you help me asses if this behavior of creating a new `AD_HOC_SUB_PROCESS_INNER_INSTANCE` with activate modification is intended or a side effect?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 2]
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
...
+
+    final var processInstanceKey =
+        ENGINE
+            .processInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariables(Map.of("activateElements", List.of("A", "A")))
+            .create();
+
+    // Get all multi-instance ad-hoc subprocess instances
+    final var miInnerInstances =
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: 🔧This name implies it's the inner instance of the multi-instance, but it's not. It is the inner instance of the ad hoc sub process instance.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 3]
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
...
+    assertThat(aTasks).hasSize(2);
+
+    // when
+    // terminate task A in the first ad-hoc instance and activate task B there
+    ENGINE
+        .processInstance()
+        .withInstanceKey(processInstanceKey)
+        .modification()
+        .terminateElement(aTasks.getFirst().getKey())
+        .activateElement("B", miInnerInstances.getFirst().getKey())
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: 🔧Here you are choosing the first ad hoc sub process' inner instance, assuming that it's the same as the flow scope of the task instance that you're terminating. Luckily that's the case, because the ordering is like that on the log. However, it would be better to be more explicit about the activation. What we want to do is move A to B within it's flowscope. Like you say in the comment "terminate task A in the first ad-hoc instance and activate task B there", so let's choose to activate B inside the same flowscope explicitly:
```suggestion
        .activateElement("B", aTasks.getFirst().getValue().getFlowScopeKey())
```

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 4]
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
...
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+                .withProcessInstanceKey(processInstanceKey)
+                .withElementId("B"))
+        .describedAs("Expect one Task B to be activated in the first ad-hoc subprocess instance")
+        .hasSize(1)
+        .extracting(Record::getValue)
+        .extracting(ProcessInstanceRecordValue::getFlowScopeKey)
+        .describedAs("Expect B Task to be activated in the specific ad-hoc subprocess instance")
+        .contains(miInnerInstances.getFirst().getKey());
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: 🔧Here we can argue the same as before. We want to assert that Task B was activated in the same ad-hoc subprocess inner instance as Task A. So let's just check that the flowscope is the same as the flow scope of Task A. 

Taking this step in addition to my previous suggestion removes the need for the miInnerInstances variable, which means one fewer RecordingExporter usage. So the test would be slightly faster as well.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java (hunk 1 above)
Comment (korthout) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: ❌Let's add a test case for this. It's useful to make this behavior clear in our tests. We can then reflect the chosen behavior if we choose to keep it the same or if we change it.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 5]
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
...
+  }
+
+  @Test
+  public void shouldModifyElementsInMultiInstanceAdHocSubprocessWithAiAgentTaskPattern() {
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: 💭I'm not sure how much this test adds. The case is not different from the one before, except that you activate two elements. The service task to determine the selection has no impact on this test case. It might be useful to remove that part from it to reduce the blast-radius. But even then, I wonder how much value this case adds to the previous. Perhaps if it would target different ancestor scopes (one targets an ad hoc subprocess inner instance, and the other target the ad hoc subprocess itself).

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 6]
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
...
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_TERMINATED)
+                .withRecordKey(aTasks.getFirst().getKey())
+                .exists())
+        .describedAs("Expect first A Task to be terminated")
+        .isTrue();
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+                .withProcessInstanceKey(processInstanceKey)
+                .withElementId("B"))
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: :x: This RecordingExporter usage is not short-circuited causing the test to take at least 5 seconds.

**Review by korthout — CHANGES_REQUESTED** [review]
Nice work @mustafadagher 👏 

🔧 I have a few suggestions, and ❌ two change requests.

Please have a look.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java (hunk 5 above)
Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: In terms of PI modification, it serves no distinctive purpose indeed, and can be considered redundant.
I was just validating one of the use cases how we use ad-hoc subprocesses for Agentic AI. I shall remove it.

>Perhaps if it would target different ancestor scopes (one targets an ad hoc subprocess inner instance, and the other target the ad hoc subprocess itself).

I'm not completely sure I fully get the comment above.

Comment (korthout) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: >Perhaps if it would target different ancestor scopes (one targets an ad hoc subprocess inner instance, and the other target the ad hoc subprocess itself).

That might be a different edge case

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java (hunk 4 above)
Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: >Taking this step in addition to my previous suggestion removes the need for the miInnerInstances variable, which means one fewer RecordingExporter usage. So the test would be slightly faster as well.

I actually refactored the above test that only relied on querying the flowscope key of the task to instead use the mi-instance key, to make the use case more clear.
I'm not concerned personally about the few milliseconds the extra RecordingExporter query adds, trading off clarity for it.

I'll add an extra assert that verifies that task-a flowscope is the same as the first adhoc-inner-instance key.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java (hunk 1 above)
Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: added a new test case to cover this scenario in https://github.com/camunda/camunda/pull/40453/commits/2b50d4579899b943af776d9ed97ecbe2461097d3

**Comment by mustafadagher** [comment]
@korthout I applied some of the suggestions, and added a new test case for the `ad-hoc inner instance` creation. Could you please review again?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java (hunk 1 above)
Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: @saig0 since @remcowesterhoud is out sick, maybe you can also fill in here if you have any thoughts about the above?

Comment (saig0) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: > creating a new AD_HOC_SUB_PROCESS_INNER_INSTANCE with activate modification is intended or a side effect?

Intended. The `AD_HOC_SUB_PROCESS` contains one `AD_HOC_SUB_PROCESS_INNER_INSTANCE` for each activated element. The inner instance acts as a variable scope. 

Conceptually, you can think about a multi-instance subprocess with the activated elements as the input collection. The `AD_HOC_SUB_PROCESS` = `MULTI_INSTANCE_BODY` and `AD_HOC_SUB_PROCESS_INNER_INSTANCE` = inner `SUB_PROCESS`.

Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java]: I see! Thanks @saig0 !
The only difference is then when you have a multi-instance ad-hoc subprocess, then you get an extra layer of `MULTI_INSTANCE_BODY` body on top of that.
If it's intended, then it's fine. It still works as intended here

**Review by korthout — APPROVED** [review]
Thanks for the changes @mustafadagher 🚀 

LGTM 👍
